### ISSUE TITLE
Per-epoch processing

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#108]
## !! WAITING ON RELEASE CANDIDATE !!

This issue is presently on-hold until the specification is stable.

## Description

This is a "mega issue" which contains multiple discreet tasks. Mega issues exist to avoid cluttering the issues page a multitude of small tasks.

If you wish to work on one of these tasks, comment below and a maintainer will break the task out into a separate issue.

## Per-epoch processing

Every `EPOCH_LENGTH` a series of functions run that mutate the validator set, perform FFG finalization and justification, process rewards/penalites and a bunch of other tasks.

## Mega issue tasks:

There is a huge amount of helpers for this component, I avoid going into detail on them here as this issue is ultimately awaiting the release candidate. Below I will outline the core functions.

- Process PoW receipt roots.
- Update justification.
- Update finalization.
- Process crosslinks.
- Process justification and finalization rewards and penalties.
- Process crosslink rewards and penalities.
- Update the validation registry.
- Reshuffle the shard proposer set.

**Comment by paulhauner** [comment]
Closing in favor of #182
